adam/experiment/live_demo.py

```diff
@@ -0,0 +1,319 @@
+                / "experiments"
+                / args.curriculum_name
+                / "test_curriculums"
+                / args.curriculum_name
+            ),
+            "file_name": "post_decode",
+            "calculate_accuracy_by_language": True,
+            "calculate_overall_accuracy": True,
+        }
+    )
+    observer = YAMLLogger.from_params("test_observer", observer_args)  # type: ignore
+
+    # Configure our 'curriculum' handling to wrap relevant functionality
+    faux_train_curriculum = DemoCurriculum(
+        args.curriculum_name, TRAINING_CURRICULUM_DIR / args.curriculum_name
+    )
+    faux_train_curriculum.initialize_curriculum()
+    curriculum = DemoCurriculum(
+        args.curriculum_name, TESTING_CURRICULUM_DIR / args.curriculum_name
+    )
+    curriculum.initialize_curriculum()
+
+    # Now we're ready to handle user input
+    while True:
+        user_input = input(MENU_STR).lower()
+        if user_input == "e":
+            print("Exiting Live Demo")
+            break
+        elif user_input == "d":
+            # Loop to handle acquiring a file input from the user.
+            while True:
+                file_path_str = input(DECODE_STR)
+                file_path = Path(file_path_str)
+                if not file_path.exists() and not file_path.is_file():
+                    print("File does not exist. Please try again.")
+                    continue
+                break
+            try:
+                # Now we need to preprocess the file, this configures the curriculum scene on disk
+                scene_dir = curriculum.preprocess(
+                    file_path, args.segmentation_api, stroke_model_path, stroke_python_env
+                )
+
+                # Now we need to load this scene into the correct perception format
+                (
+                    situation,
+                    linguistic_description,
+                    perception_frame,
+                ) = phase3_process_scene_dir(
+                    scene_dir, curriculum_type="testing", color_is_rgb=args.color_is_rgb
+                )
+
+                # Now we describe the given input
+                descriptions_from_learner = learner.describe(perception_frame)
+
+                # The observer should run now
+                if not observer:
+                    raise RuntimeError(f"Missing observer from arguments {observer_args}")
+                observer.observe(
+                    situation,
+                    linguistic_description,
+                    perception_frame,
+                    descriptions_from_learner,
+                )
+                observer.report()
+
+                print(f"Demo scene {curriculum.situation_number} has been processed.")
+            # We purposely catch broad exceptions here because we don't want the interactive session to end
+            # If an exception occurs in processing a given file
+            except Exception as e:  # pylint: disable=broad-except
```

Comment:
Not sure where to put this, but I ran into a problem while debugging before the demo -- is this still a thing or is it handled? The problem was that if preprocessing fails, then the demo curriculum path would be left in an inconsistent state where we have say 1 curriculum dir but no `info.yaml`, or 3 and the info file says 2.

Before we just crashed when this happened which made the issue more dire. When I restarted the script it wouldn't be able to process further images, because it would try to write to a directory that shouldn't exist but does (e.g. `situation_0`). I had to delete the directory and restart, because `(r)eset` didn't work at the time (the directory was no longer empty).

I'm not sure if this is still a problem now that we're catching exceptions. It's possible that because we're maintaining the curriculum state things might be okay. The same problem might come up though if you quit and restart after preprocessing crashes?

Reply:
The goal was that catching exceptions should keep the curriculum manager from going out-of-sync with disk so when we go to process a new image everything works fine. I didn't thoroughly test this but it did appear to work in theory.